File: novascript/src/runtime/native/sys_module.ts

import * as os from "os";
import { execSync } from "child_process";
import { RuntimeVal, MK_NATIVE_FN, MK_STRING, MK_BOOL, MK_NULL, MK_NUMBER, MK_OBJECT } from "../values";

export function createSysModule(version: string) {
    const sysProps = new Map<string, RuntimeVal>();
    
    sysProps.set("platform", MK_STRING(os.platform()));
    sysProps.set("arch", MK_STRING(os.arch()));
    sysProps.set("version", MK_STRING(version));
    sysProps.set("env", MK_NATIVE_FN((args) => {
        const key = (args[0] as any).value;
        const val = process.env[key];
        return val !== undefined ? MK_STRING(val) : MK_NULL();
    }));
    sysProps.set("uptime", MK_NATIVE_FN(() => MK_NUMBER(os.uptime())));
    sysProps.set("hostname", MK_NATIVE_FN(() => MK_STRING(os.hostname())));
    sysProps.set("totalmem", MK_NATIVE_FN(() => MK_NUMBER(os.totalmem())));
    sysProps.set("freemem", MK_NATIVE_FN(() => MK_NUMBER(os.freemem())));
    sysProps.set("exec", MK_NATIVE_FN((args) => {
        try {
            const output = execSync((args[0] as any).value).toString();
            return MK_STRING(output);
        } catch (e: any) {
            return MK_STRING(e.message);
        }
    }));
    sysProps.set("exit", MK_NATIVE_FN((args) => {
        process.exit((args[0] as any)?.value || 0);
    }));
    sysProps.set("time", MK_NATIVE_FN(() => MK_NUMBER(Date.now())));
    
    return MK_OBJECT(sysProps);
}

File: runtime/native/sys_module.go
package native

import (
	"os"
	"os/exec"
	"runtime"
	"time"

	nruntime "novago/runtime"
)

func CreateSysModule(version string) *nruntime.ObjectVal {
	sysProps := make(map[string]nruntime.RuntimeVal)

	sysProps["platform"] = nruntime.MK_STRING(runtime.GOOS)
	sysProps["arch"] = nruntime.MK_STRING(runtime.GOARCH)
	sysProps["version"] = nruntime.MK_STRING(version)

	sysProps["env"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		key := args[0].(*nruntime.StringVal).Value
		val, exists := os.LookupEnv(key)
		if exists {
			return nruntime.MK_STRING(val)
		}
		return nruntime.MK_NULL()
	})

	sysProps["hostname"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		name, err := os.Hostname()
		if err != nil {
			return nruntime.MK_NULL()
		}
		return nruntime.MK_STRING(name)
	})

	sysProps["exec"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		cmdStr := args[0].(*nruntime.StringVal).Value
		var cmd *exec.Cmd
		if runtime.GOOS == "windows" {
			cmd = exec.Command("cmd", "/c", cmdStr)
		} else {
			cmd = exec.Command("sh", "-c", cmdStr)
		}
		out, err := cmd.CombinedOutput()
		if err != nil {
			return nruntime.MK_STRING(string(out) + "\n" + err.Error())
		}
		return nruntime.MK_STRING(string(out))
	})

	sysProps["exit"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		code := 0
		if len(args) > 0 {
			if num, ok := args[0].(*nruntime.NumberVal); ok {
				code = int(num.Value)
			}
		}
		os.Exit(code)
		return nruntime.MK_NULL()
	})

	sysProps["time"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		return nruntime.MK_NUMBER(float64(time.Now().UnixMilli()))
	})

	return nruntime.MK_OBJECT(sysProps)
}
